File: Python/Treap.py
from random import randrange as rand

class Node:
	def __init__(self, x):
		self.x = x
		self.y = rand(1000000000)
		self.l = self.r = None

class Treap:
	def __init__(self):
		self.root = None

	# join two trees O(log n)
	def join(self, l, r):
		if not l: return r
		if not r: return l

		if l.y < r.y:
			l.r = self.join(l.r, r)
			return l
		else:
			r.l = self.join(l, r.l)
			return r

	# split into two trees O(log n)
	def split(self, x, p):
		if not p:
			return (None, None)

		if x < p.x:
			l, r = self.split(x, p.l)
			p.l = r
			return (l, p);
		else:
			l, r = self.split(x, p.r)
			p.r = l
			return (p, r)

	# O(log n)
	def insert(self, x):
		l, r = self.split(x, self.root)
		self.root = self.join(l, self.join(Node(x), r))

	# delete one element O(log n)
	def delete(self, x):
		l, r = self.split(x, self.root)
		l2, r2 = self.split(x-1, l)
		l = self.join(l2, self.join(r2.l, r2.r) if r2 else None)
		self.root = self.join(l, r)

	# delete range of elements O(log n)
	def delete(self, L, R):
		l, r = self.split(R, self.root)
		l2, r2 = self.split(L-1, l)
		self.root = self.join(l2, r)

	# O(log n)
	def find(self, x):
		p = self.root
		while p:
			if x == p.x:
				return True
			
			if x < p.x:
				p = p.l
			else:
				p = p.r
		return False

	# O(n)
	def inorder(self, p, level, view):
		if not p: return level
		l = self.inorder(p.l, level + 1, view)
		if view: print(p.x)
		r = self.inorder(p.r, level + 1, view)
		return max(l, r)

	# O(n)
	def disply(self, view = True):
		return self.inorder(self.root, 0, view)

t = Treap()
for i in range(10000):
	t.insert(i)

t.delete(6, 9994)
print("Levels: %d" % t.disply())

File: Kotlin/Treap.kt
class Treap {
    data class Node(var x: Int,
                    var l: Node? = null, var r: Node? = null,
                    var s: Int = 1, val y: Int = (Math.random() * 100_000).toInt())
    
    fun Node.upd(): Node {
        s = 1
        s += l?.s ?: 0
        s += r?.s ?: 0
        return this
	}
    
    var root: Node? = null
    
    fun join(l: Node?, r: Node?) : Node? {
        if(l == null) return r
        if(r == null) return l
        
        if(l.y < r.y) {
        	l.r = join(l.r, r)
            return l.upd()
        } else {
        	r.l = join(l, r.l)
            return r.upd()
		}
	}
    
    fun split(p: Node?, x: Int): Pair<Node?, Node?> {
        if(p == null)
        	return Pair(null, null)
        
        if(x < p.x) {
        	var (l, r) = split(p.l, x)
            p.l = r
            r = p.upd()
            return Pair(l, r)
        } else {
        	var (l, r) = split(p.r, x)
            p.r = l
            l = p.upd()
            return Pair(l, r)
		}
	}
    
    fun insert(x: Int) {
        val (l, r) = split(root, x)
        root = join(l, join(Node(x), r))
	}
    
    fun remove(x: Int) {
        root = remove(root, x)
	}
    
    fun print(): Int {
        return print(root)
	}
    
    fun remove(p: Node?, x: Int): Node? {
        if(p == null)
        	return null
        
        if(x == p.x)
        	return join(p.l, p.r)
        
        if(x < p.x)
        	p.l = remove(p.l, x)
        else
        	p.r = remove(p.r, x)
        
        return p.upd()
	}
    
    fun print(p: Node?): Int {
        if(p == null)
        	return 0
        
        val l = print(p.l)
        println(p.x)        
        val r = print(p.r)
        
        return 1 + intArrayOf(l, r).max()!!
    }
}

fun main(args: Array<String>) {
    val n = 100
    
    val t = Treap()
    for(i in 0 until n)
    	t.insert(i)
    
    t.print()
}
